### PR TITLE
*: replace functions in intarith.h with the ones provided by standard library

### PR DESCRIPTION
**Pull request by tchaikov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
```
2022-08-21 18:33:20,836.836 INFO:__main__:test_all (tasks.mgr.dashboard.test_rgw.RgwBucketTest) ... ERROR
2022-08-21 18:33:20,837.837 INFO:__main__:
2022-08-21 18:33:20,837.837 INFO:__main__:======================================================================
2022-08-21 18:33:20,837.837 INFO:__main__:ERROR: test_all (tasks.mgr.dashboard.test_rgw.RgwBucketTest)
2022-08-21 18:33:20,837.837 INFO:__main__:
2022-08-21 18:33:20,837.837 INFO:__main__:----------------------------------------------------------------------
2022-08-21 18:33:20,837.837 INFO:__main__:Traceback (most recent call last):
2022-08-21 18:33:20,837.837 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/mgr/dashboard/helper.py", line 271, in setUp
2022-08-21 18:33:20,837.837 INFO:__main__:    self.wait_for_health_clear(self.TIMEOUT_HEALTH_CLEAR)
2022-08-21 18:33:20,837.837 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/ceph_test_case.py", line 175, in wait_for_health_clear
2022-08-21 18:33:20,837.837 INFO:__main__:    self.wait_until_true(is_clear, timeout)
2022-08-21 18:33:20,837.837 INFO:__main__:  File "/home/jenkins-build/build/workspace/ceph-api/qa/tasks/ceph_test_case.py", line 212, in wait_until_true
2022-08-21 18:33:20,837.837 INFO:__main__:    raise TestTimeoutError("Timed out after {0}s and {1} retries".format(elapsed, retry_count))
2022-08-21 18:33:20,838.838 INFO:__main__:tasks.ceph_test_case.TestTimeoutError: Timed out after 60s and 0 retries
```

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Review by Matan-B — COMMENTED** [review]
Should we also use std::popcount instead of intaith.h one?

**Comment by tchaikov** [comment]
> Should we also use std::popcount instead of intaith.h one?

yeah, actually, i do have two commits in my local branch to do this change. but it turns out one of my changes fails the test. still bisecting it.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/os/bluestore/BlueStore.h [hunk 1]
```diff
@@ -3836,7 +3837,7 @@ class BlueStoreRepairer
 	      uint64_t min_alloc_size,
 	      uint64_t mem_cap = DEF_MEM_CAP) {
       ceph_assert(!granularity); // not initialized yet
-      ceph_assert(min_alloc_size && isp2(min_alloc_size));
+      ceph_assert(min_alloc_size && std::has_single_bit(min_alloc_size));
```
Comment (idryomov) [inline src/os/bluestore/BlueStore.h:3840]: `std::has_single_bit` already takes care of 0 -- `isp2` should have been named `isp2_or_zero`:
```suggestion
      ceph_assert(std::has_single_bit(min_alloc_size));
```

Comment (tchaikov) [reply src/os/bluestore/BlueStore.h]: @idryomov fixed and repushed.

**Comment by tchaikov** [comment]
i also have a commit to replace `ctz()` with `std::countr_zero()`, but it turns out it fails the test. i will try to fix it and include that commit in another PR.

**Review by Matan-B — APPROVED** [review]

**Review by idryomov — APPROVED** [review]

File: src/include/intarith.h [hunk 2]
```diff
@@ -150,36 +152,9 @@ template<class T>
...
-  (std::is_integral<T>::value &&
-   sizeof(T) > sizeof(unsigned long) &&
-   sizeof(T) <= sizeof(unsigned long long)),
-  unsigned>::type cbits(T v) {
-  if (v == 0)
-    return 0;
-  return (sizeof(v) * 8) - __builtin_clzll(v);
+template<std::integral T>
+unsigned cbits(T v) {
+  return (sizeof(v) * 8) - std::countl_zero(std::make_unsigned_t<T>(v));
```
Comment (rzarzynski) [inline src/include/intarith.h:123]: How about `CHAR_BITS` here? Anyway, the number already existed in the code, so this is not even a nit :-).

**Review by rzarzynski — APPROVED** [review]

**Comment by rzarzynski** [comment]
Just to the record (wanted to see how `countl_zero` is exactly implemented underneath):

```cpp
  /// The number of contiguous zero bits, starting from the highest bit.
  template<typename _Tp>
    constexpr _If_is_unsigned_integer<_Tp, int>
    countl_zero(_Tp __x) noexcept
    { return std::__countl_zero(__x); 
```

```cpp
  template<typename _Tp>
    constexpr int
    __countl_zero(_Tp __x) noexcept
    {
      using __gnu_cxx::__int_traits;
      constexpr auto _Nd = __int_traits<_Tp>::__digits;

      if (__x == 0)
        return _Nd;

      constexpr auto _Nd_ull = __int_traits<unsigned long long>::__digits;
      constexpr auto _Nd_ul = __int_traits<unsigned long>::__digits;
      constexpr auto _Nd_u = __int_traits<unsigned>::__digits;

      if _GLIBCXX17_CONSTEXPR (_Nd <= _Nd_u)
        {
          constexpr int __diff = _Nd_u - _Nd;
          return __builtin_clz(__x) - __diff;
        }
      else if _GLIBCXX17_CONSTEXPR (_Nd <= _Nd_ul)
        {
          constexpr int __diff = _Nd_ul - _Nd;
          return __builtin_clzl(__x) - __diff;
        }
      else if _GLIBCXX17_CONSTEXPR (_Nd <= _Nd_ull)
        {
          constexpr int __diff = _Nd_ull - _Nd;
          return __builtin_clzll(__x) - __diff;
        }
      else // (_Nd > _Nd_ull)
        {
          static_assert(_Nd <= (2 * _Nd_ull),
                        "Maximum supported integer size is 128-bit");

          unsigned long long __high = __x >> _Nd_ull;
          if (__high != 0)
            {
              constexpr int __diff = (2 * _Nd_ull) - _Nd;
              return __builtin_clzll(__high) - __diff;
            }
          constexpr auto __max_ull = __int_traits<unsigned long long>::__max;
          unsigned long long __low = __x & __max_ull;
          return (_Nd - _Nd_ull) + __builtin_clzll(__low);
        }
    }
```

**Comment by tchaikov** [comment]
jenkins test api

File: src/include/intarith.h (hunk 2 above)
Comment (tchaikov) [reply src/include/intarith.h:123]: good idea. less magic numbers then! i will do this in the promised follow-up PR.

**Pull request merged**